[RISCV] Align MCOperandPredicates with AsmParser

lenary opened the pull request:
Before this change, many MCOperandPredicates accepted bare symbols but there was no cohesive logic as to which did and which didn't, and it did not correspond to where the parser did and did not accept bare symbols.

MCOperandPredicates are used by the assembly printer when choosing to whether to use an alias, and by the assembler compression and decompression mechanism. RISC-V uses both of these extensively.

Xqci had a bug with a compress pattern for `qc.e.li` (to `c.li`) which was causing crashes because `RISCVMCCodeEmitter::getImmOpValue` could not handle a bare symbol ref when encoding the `c.li` instruction.

This patch aligns the MCOperandPredicates with the predicates in the RISC-V assembly parser, denying bare symbol references in places they were accepted for aliases/compression in the past.

This is not NFC, as some assembly printing has changed with the `li` alias for `addi`, but this seems very minor. The crash when assembling `qc.e.li a0, undef` is now fixed.

svs-quic commented on llvm/test/MC/RISCV/xqcibi-relocations.s:
```diff
@@ -86,6 +86,22 @@ qc.e.bgeui s2, 24, same_section
 
 .option norelax
```
Probably dont need this and the option relax below?

lenary replied:
Done

svs-quic commented on llvm/test/MC/RISCV/xqcilb-relocations.s:
```diff
@@ -94,6 +94,24 @@ qc.e.jal same_section
 
 .option norelax
```
Same here.

lenary commented:
I've become more convinced that this change is "correct" by examining the behaviour in the assembler. Broadly, it's worth noting that `isSImm6` in the assembler does not accept any symbols, so this aligns the `MCOperandPredicate` with that. This does not affect absolute symbol handling, because the parser turns absolute symbols into immediates as soon as it can, from what i can see with `-show-inst` (this doesn't apply to absolute symbols defined after the instruction is parsed, but these error at parse-time like an undefined symbol anyway)

I will revisit the structure of this PR:
- 1 NFC (pre-)commit adding tests for instructions with absolute, relative, undef, and latedef (an absolute symbol def made after the symbol is used) symbols.
- 1 commit to align the MCOperandPredicates with the parser predicates.

The second commit, if i get this right, should only change the behaviour for the `qc.e.li a0, undef` case which is currently crashing.

lenary committed:
[RISCV][NFC] Add Coverage For Symbol Operands

This adds assembly test coverage for the following operand types:
- simm12
- uimm20_lui
- uimm20_auipc
- simm6
- simm6nonzero
- simm5
- simm5_plus1
- simm20_li

lenary committed:
[RISCV] Align MCOperandPredicates with AsmParser

Before this change, many MCOperandPredicates accepted bare symbols but
there was no cohesive logic as to which did and which didn't, and it did
not correspond to where the parser did and did not accept bare symbols.

MCOperandPredicates are used by the assembly printer when choosing to
whether to use an alias, and by the assembler compression and
decompression mechanism. RISC-V uses both of these extensively.

Xqci had a bug with a compress pattern for `qc.e.li` (to `c.li`) which
was causing crashes because `RISCVMCCodeEmitter::getImmOpValue` could
not handle a bare symbol ref when encoding the `c.li` instruction.

This patch aligns the MCOperandPredicates with the predicates in the
RISC-V assembly parser, denying bare symbol references in places they
were accepted for aliases/compression in the past.

This is not NFC, as some assembly printing has changed with the `li`
alias for `addi`, but this seems very minor. The crash when assembling
`qc.e.li a0, undef` is now fixed.

svs-quic commented on llvm/lib/MC/MCInst.cpp:
```diff
@@ -63,12 +63,29 @@ bool MCOperand::evaluateAsConstantImm(int64_t &Imm) const {
+}
+
+bool MCOperand::isSimpleSymbolRef(MCExpr::Spec &Specifier) const {
+  assert(isExpr() && "isBareSymbolRef expects only expressions");
```
Do you maybe want to change this to isSimpleSymbolRef?

svs-quic reviewed (approved):
The Xqci specific changes look good to me. Please wait for someone else to review the rest.

lenary committed:
Address feedback, comment on target kind

lenary commented on llvm/lib/MC/MCInst.cpp:
```diff
@@ -63,12 +63,31 @@ bool MCOperand::evaluateAsConstantImm(int64_t &Imm) const {
+
+bool MCOperand::isSimpleSymbolRef(MCExpr::Spec &Specifier) const {
+  assert(isExpr() && "isSimpleSymbolRef expects only expressions");
   const MCExpr *Expr = getExpr();
   MCExpr::ExprKind Kind = getExpr()->getKind();
-  return Kind == MCExpr::SymbolRef &&
-         cast<MCSymbolRefExpr>(Expr)->getSpecifier() == 0;
+
+  switch (Kind) {
+  case MCExpr::Binary:
+  case MCExpr::Unary:
+  case MCExpr::Constant:
+    break;
+  case MCExpr::Target:
+    // It's not clear this is right, does MCTargetExpr need another virtual method?
+    break;
+  case MCExpr::SymbolRef:
+    Specifier = cast<MCSymbolRefExpr>(Expr)->getSpecifier();
+    return true;
+  case MCExpr::Specifier:
+    Specifier = cast<MCSpecifierExpr>(Expr)->getSpecifier();
+    return true;
+  }
+  return false;
```
@MaskRay it would be good to understand your view on this method - the aim is to effectively capture everywhere which is just a specifier and one symbol, extracting the specifier - like what `RISCVAsmParser::classifySymbolRef` does.

I haven't had to cover MCTargetExpr because only RISC-V uses this method right now, and that backend now uses MCSpecifierExpr. I guess maybe I should be using `Expr->evaluateAsRelocatable(...))` and looking at the resulting `MCValue`, rather than adding a new virtual method?

MaskRay replied:
Looking. I have migrated most targets to use MCSpecifierExpr.

lenary committed:
clang-format

MaskRay commented:
The introduction of `MCOperand::isBareSymbolRef` in https://reviews.llvm.org/D45385 (2018) seems problematic.
Generally, the assembly parser should treat expressions like `sym, sym + 0, or (sym + 3) - 3` equivalently. Most targets don't have the "bare symbol" notation.

The use of `return MCOp.isBareSymbolRef();` in RISC-V TableGen files appears to be an inappropriate overuse and should be eliminated.
Unless needed to resolve parser ambiguity (as noted in a patch you co-authored, though I can't recall), `MCOperand::isBareSymbolRef` should be avoided.
I recall that the parser ambiguity is due to an LLVM AsmMatcher issue. GNU Assembler's opcodes/riscv-opc.c driven parser doesn't have the issue.

lenary commented:
ping

MaskRay commented:
Sorry for the delay. I'll take a look. The alternative approach that doesn't introduce MCExpr stuff into the MCInst interface looks promising. (The existing `MCInst::isBareSymbolRef` by [reviews.llvm.org/D45385](https://reviews.llvm.org/D45385) is a hack. If anything, I hope that we limit hacks to RISCVAsmParser instead of the generic interface...)

MaskRay reviewed (changes requested):
Left a comment in in the alternative approach.

I support the changes that eliminate some `MCOp.isBareSymbolRef()` hacks introduced in https://reviews.llvm.org/D45385 (2018). Thanks!
That said, I clicked "request changes" as I strongly recommend that RISC-V avoids adding more hacks to the generic interface like MCInst `MCOperand::isSimpleSymbolRef`.
AsmParser should not check the shape of a `MCExpr`.

lenary commented:
I think, given your comments, that you want me to go with #146763, which I have updated.

I am going to push back on one of your comments here slightly:
> AsmParser should not check the shape of a `MCExpr`.

I agree that what's currently happening in the parser is wrong, as we should be accepting bare symbols and other expressions in a lot more places (because they might be absolute values). That said, I think we can and should exclude expressions with invalid specifiers in the parser. For instance, if I write `beq a0, a1, %pcrel_lo(<anything>)`, we know that this is wrong straightaway.

As I've said before, I'm happy to work towards making the assembly parser more lenient in the future, BUT I doing so for LLVM 21 is unsustainable this late in the release cycle - but the compression bug with `qc.e.li` can and should be fixed this release cycle.

MaskRay commented:
The changes that remove MCOp.isBareSymbolRef() uses look good, e.g. the change that replaces

```
def simm5 : RISCVSImmLeafOp<5> {
  let MCOperandPredicate = [{
    int64_t Imm;
    if (MCOp.evaluateAsConstantImm(Imm))
      return isInt<5>(Imm);
    return MCOp.isBareSymbolRef();
  }];
}
```

with

```
def simm5 : RISCVSImmLeafOp<5> {
  let MCOperandPredicate = [{
    int64_t Imm;
    if (!MCOp.evaluateAsConstantImm(Imm))
      return false;
    return isInt<5>(Imm);
  }];
}
```

We might want an API to extract the relocation specifier from a MCOperand, but it should not be called `SymbolRef` as `%specifier(sym_a-sym_b+cst)` should be allowed as well.


I just realized that this is closely tied to relocatin specifier design issue I filed in April: https://github.com/llvm/llvm-project/issues/135592 .
I'm not satisfied with how LLVM MC handles relocation specifier encoding.
Relocation specifier can be encoded by every sub-expression in the MCExpr tree, rather than the fixup itself (or the instruction, as in GNU Assembler).

```
MCBinaryExpr
  MCSymbolRef specifier
  MCSymbolRef

MCBinaryExpr
  MCSymbolRef
  MCSymbolRef specifier

RISCVMCExpr specifier
  MCSymbolRef

AArch64MCExpr specifier
  MCSymbolRef

// now XXXMCExpr have been unified to MCSpecifierExpr
```

It might be ideal to encode the MCExpr and relocation specifier as a pair, eliminating the need for MCSpecifierExpr to simply wrap a pointer, which would save memory.
However, MCOperand is size-sensitive, and we cannot include an additional integer as part of MCOperand.

We should consider the interplay of MCFixup and MCOperand, and how to encode relocation specifier. A change like this that excludes certain relocation specifiers from MCOperandPredicate, can be valuable in the future.
